src/Analyser/NodeScopeResolver.php

```diff
@@ -3193,12 +3225,14 @@ static function (): void {
 				$hasYield = $hasYield || $result->hasYield();
 				$throwPoints = array_merge($throwPoints, $result->getThrowPoints());
 				$impurePoints = array_merge($impurePoints, $result->getImpurePoints());
+				$isAlwaysTerminating = $isAlwaysTerminating || $result->isAlwaysTerminating();
 				$scope = $result->getScope();
 			}
 		} elseif ($expr instanceof ArrayDimFetch) {
 			$hasYield = false;
 			$throwPoints = [];
 			$impurePoints = [];
+			$isAlwaysTerminating = false;
```

Comment:
Basically this should be updated every time `$hasYield` is updated. Please review all branches in `processExprNode` like that.

Reply:
I adjusted it where it made sense to me.

I left out e.g. null-safe operations, as these would only terminate for always-non-nullables and similar stuff.